gitpod-io/gitpod
# [content-service] Add support to use existing S3 bucket

## Description
This PR introduces a new "single bucket mode" to the S3 storage implementation.

In some scenarios, we could not have enough permissions to create buckets. In those cases the "one-bucket per user" implementation fails.

If there's a bucket name configured (see installer/content-service config change), that bucket name is used and the user ID is prefixed to the object name. If no such bucket name is configured, the old "one bucket per user" behaviour kicks in. This way, users can maintain backwards compatibility for now. There is no automatic migration or "hybrid mode" where we fall back to the old behaviour. We don't have enough installations in the wild which were able to actually use the old behaviour to justify introducing such a mode.

We maintain the old mode in GCP (and maybe even S3) for now because 
- it maintains backwards compatibility
- it allows for region-local optimisations. That said, we could just as well introduce single regional buckets instead.
- moving away from it would require migration for gitpod.io

## How to test
Precursor: ensure the installation uses minio/S3, and that the `bucket` is configured
- start/stop a workspace, make sure the backup happens and is restored
- take a snapshot and open a workspace from it
- run a prebuild and open a workspace from it
- use the VS Code settings sync, make sure it still syncs
- check out prebuild logs in the dashboard

## Release Notes

```release-note
[content-service] Add support to use a single S3 bucket
```


Fixes https://github.com/gitpod-io/gitpod/issues/10070
